Tools transition prepare release

gurkanindibay opened the pull request.

onurctirtir commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+        lines = file_content.splitlines()
+        line_counter = 0
+        has_match = False
+        for line in lines:
+            if re.match(match_regex, line):
+                has_match = True
+                lines[line_counter] = replace_str
+            line_counter = line_counter + 1
```
In general, we can use `enumerate` instead of doing `range(len(iterable))` or keeping an counter

```suggestion
        for line_number, line in enumerate(lines):
            if re.match(match_regex, line):
                has_match = True
                lines[line_number] = replace_str
```

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/common_validations.py:
```diff
@@ -0,0 +1,28 @@
+
+@parameter_validation
+def is_version(version: str):
+    if version is None or not version:
```
can't we simply check `if not version:` ?

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/common_validations.py:
```diff
@@ -0,0 +1,28 @@
+
+@parameter_validation
+def is_tag(tag: str):
+    if tag is None or not tag:
```
can't we simply check `if not tag:` ?

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/common_validations.py:
```diff
@@ -0,0 +1,28 @@
+        raise ValueError("version should be non-empty and should not be None")
+    if not re.match(r"\d+\.\d+\.\d+$", version):
+        raise ValueError(
+            "version should include 3 levels of versions consists of numbers  separated with dots. e.g: 10.0.1")
```
better wording: "version should include three level of digits separated by dots, e.g: 10.0.1"

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/common_validations.py:
```diff
@@ -0,0 +1,28 @@
+        raise ValueError("tag  should be non-empty and should not be None")
+    if not re.match(r"v\d+\.\d+\.\d+$", tag):
+        raise ValueError(
+            "tag  should start with v and should include 3 levels of versions consists of numbers " +
+            "separated with dots. e.g: v10.0.1")
```
better wording: "tag should start with 'v' and should include three level of digits separated by dots, e.g: v10.0.1"

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,265 @@
+CONFIGURE = "configure"
+CITUS_CONTROL_SEARCH_PATTERN = r"^default_version*"
+
+MULT_EXT_MAJOR_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}devel$"
+MULT_EXT_PATCH_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}.\d{1,2}$"
```
1) MULT -> MULTI
2) MULT_EXT_PATCH_SEARCH_PATTERN -> MULT_EXT_MINOR_PATCH_SEARCH_PATTERN, since that pattern seems to be used for both minor & patch releases

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/common_validations.py:
```diff
@@ -0,0 +1,28 @@
+def is_version(version: str):
+    if version is None or not version:
+        raise ValueError("version should be non-empty and should not be None")
+    if not re.match(r"\d+\.\d+\.\d+$", version):
```
I think we can define a pattern like `CITUS_VERSION_PATTERN = "\d{1,2}\.\d{1,2}.\d{1,2}"` and use it in `is_version`, `is_tag` and in `MULT_EXT_x_SEARCH_PATTERN` vars

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,265 @@
+                               f'{str(int(project_version_details["patch"]) + 1)}'
+    upcoming_version = default_upcoming_version if os.getenv("UPCOMING_VERSION") is None else os.getenv(
+        "UPCOMING_VERSION")
+    upcoming_version_details = get_version_details(upcoming_version)
+    upcoming_patch_version = f'{upcoming_version_details["major"]}.{upcoming_version_details["minor"]}'
```
1) upcoming_patch_version -> upcoming_minor_version

2) it seems that rest of the code doesn't need to know about `upcoming_version_details`, can we move those two lines into a function like `get_upcoming_minor_version`

In general, for the places in `update_release` that we are building strings, can we have some nice short functions to improve the readability ?

------

I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments.

gurkanindibay replied:
"In general, for the places in update_release that we are building strings, can we have some nice short functions to improve the readability ?"
I think transforming single line operations into methods hurts readibility. Because single line logic will be elsewhere and to understand the operation you shoould go that tline

"I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments."
I transformed code into series of methods as you suggest

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,265 @@
+        run(f"git checkout {main_branch}")
+        run(f"git pull")
+        print(f"### OK {main_branch} checked out and pulled### ")
+        # create release branch in release-X.Y format
+        print(f"### Creating release branch with name {release_branch_name}...### ")
+        run(f'git checkout -b {release_branch_name}')
+        print(f"### OK {release_branch_name} created### ")
+        resource_status = ResourceStatus.RELEASE_BRANCH_LOCAL
```
>I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments.

e.g 1

gurkanindibay replied:
There is not much logic in this method If I group and make method for each of them, readability will be hurt. Most of them are 3 lines and two of them are releated to logging start and finish

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,265 @@
+                                    f"ALTER EXTENSION citus UPDATE TO '{upcoming_patch_version}-1';"):
+            raise ValueError(f"{multi_extension_out_path} does not have match for version")
+        print(f"### OK Current schema version updated on {multi_extension_out_path} to {upcoming_patch_version}### ")
+        # create new sql file with the name ">./src/backend/distributed/citus--$current_schema_version--
+        # $upcoming_minor_version-1.sql"
+        newly_created_sql_file = f"citus--{current_schema_version}--{upcoming_patch_version}-1.sql"
+        print(f"### Creating file {newly_created_sql_file}...### ")
+        with open(f"{distributed_dir_path}/{newly_created_sql_file}", "w") as f_writer:
+            content = f"/* citus--{current_schema_version}--{upcoming_patch_version}-1 */"
+            content = content + "\n\n"
+            content = content + f"-- bump version to {upcoming_patch_version}-1" + "\n\n"
+            f_writer.write(content)
+        run(f"git add {distributed_dir_path}/{newly_created_sql_file}")
+        print(f"### OK {newly_created_sql_file} created.")
```
>I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments.

e.g 2

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir reviewed (commented):
Sharing my comments from initial round.
Please let me know once you address the comments, so that I can do a more detailed review for `prepare_release.py` and for two `_test_.py` files

In general, please try to move the things in `update_release` into smaller logical units.
That way, I (and the other reviewers/readers) can easily understand the code & focus more on the end-2-end behavior.

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+    return version_info["patch"] == "0"
+
+
+def str_array_to_str(str_array: List[str]) -> str:
+    result_str = ""
+    for i in range(len(str_array)):
+        result_str = result_str + str_array[i] + "\n"
+    return result_str
```
A one more new line should be added to make them equal. I'm fixing it

gurkanindibay replied:
Fixed

gurkanindibay replied:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+    filtered_prs = []
+    for pr in prs:
+        filtered_labels = filter(lambda label: label.name == label_name, pr.labels)
+        if len(list(filtered_labels)) > 0:
```
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+            break
+        if pull.merged_at is not None and pull.merged_at > earliest_date:
+            picked_pulls.append(pull)
+    return picked_pulls
```
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+            if line == line_content:
+                found = True
+                break
+    return found
```
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,151 @@
+            if line == line_content:
+                counter = counter + 1
+
+    return counter
```
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+CONFIGURE = "configure"
+CITUS_CONTROL_SEARCH_PATTERN = r"^default_version*"
+
+MULTI_EXT_MAJOR_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}devel$"
+MULTI_EXT_MINOR_PATCH_SEARCH_PATTERN = r"^\s*" + CITUS_VERSION_PATTERN
```
Evaluation of `MULTI_EXT_MINOR_PATCH_SEARCH_PATTERN` seems little bit odd, since we close the regex with `$` in the variable that we use (`CITUS_VERSION_PATTERN`). Also left some new naming suggestions for them.

```suggestion
MULTI_EXT_DEVEL_SEARCH_PATTERN = r"^\s*" + CITUS_MINOR_VERSION_PATTERN + "devel$"
MULTI_EXT_PATCH_SEARCH_PATTERN = r"^\s*" + CITUS_PATCH_VERSION_PATTERN + r"$"

(in common_validations.py)
CITUS_PATCH_VERSION_PATTERN = CITUS_MINOR_VERSION_PATTERN + r".\d{1,2}"
CITUS_MINOR_VERSION_PATTERN = r"\d{1,2}\.\d{1,2}"
```

[means MULTI_EXT_MAJOR_SEARCH_PATTERN -> MULTI_EXT_DEVEL_SEARCH_PATTERN & MULTI_EXT_MINOR_PATCH_SEARCH_PATTERN -> MULTI_EXT_PATCH_SEARCH_PATTERN
(second one is my bad from the previous round)]

gurkanindibay replied:
Fixed

onurctirtir reviewed (commented):
1) I skipped reviewing most parts of `common_tool_methods.py`, since some of them seem to be related with #90 and it's hard to identify which ones they are. I'd prefer taking another look into those functions in #90.

2) Why do we have an empty file named as `bash/prepare_release.sh` ?

3) I compared `update_release` with `prepare_release.pl` script. At the high level it looks correct from behavioral perspective, except two comments that I shared. After you address them, I would like manually test this part before approving.

4) I plan to go over log messages, comments and test py files separately today

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        create_branch(release_pr_branch)
+        if not is_test:
+            push_branch(release_pr_branch)
+    return release_branch_name, upcoming_version_branch, f"{DISTRIBUTED_DIR_PATH}/{newly_created_sql_file}", \
+           resource_status
```
Instead of returning a tuple, can we define a class such as `UpdateReleaseReturnValue` ?

gurkanindibay replied:
Fixed

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    print(f"OK Cherry pick completed for all PRs on branch {release_branch_name}")
+
+
+def create_pull_request_for_upcoming_version_branch(main_branch, repository, upcoming_version,
+                                                    upcoming_version_branch):
```
maybe we can re-order func args: repository, `repository, main_branch, upcoming_version_branch, upcoming_version`

gurkanindibay replied:
Fixed

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    print(
+        f"### Updating schema version {current_schema_version} on {multi_extension_sql_path} "
+        f"file with the upcoming version {upcoming_version}...### ")
+    # TODO Append instead of replace may require
```
what does that mean ?

gurkanindibay replied:
CHangin schema version with the upcoming version param

onurctirtir replied:
Yes, it's the case indeed.
We should add new lines with `ALTER EXTENSION citus UPDATE TO` commands, should not replace the previous ones.

gurkanindibay replied:
When I look these lines I see that there is a replace operation but when I look comment it was supposed to be append
I'm confused. Which one that should be?
https://github.com/citusdata/tools/blob/b55e5e717faf61ca3da4a00559c1ace6f469372c/automated_packaging/prepare_release.pl#L95-L96

onurctirtir replied:
we need to append. I believe, above `sed` commands also do that

gurkanindibay replied:
Yes I've just realized that -i option appends line. I'll fix it

onurctirtir replied:
then waiting for that item to be resolved, thanks

gurkanindibay replied:
Fixed it

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    if not replace_line_in_file(citus_control_file_path, CITUS_CONTROL_SEARCH_PATTERN,
+                                f"default_version = '{upcoming_minor_version}-1'"):
+        raise ValueError(f"{citus_control_file_path} does not have match for version")
+    print(f"### OK{citus_control_file_path} file updated with the upcoming version {upcoming_minor_version}...### ")
```
```suggestion
    print(f"### OK {citus_control_file_path} file is updated with the upcoming version {upcoming_minor_version}...### ")
```

gurkanindibay replied:
Fixed

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    return current_schema_version
+
+
+def update_version_with_upcoming_version_inconfig_py(config_py_path, upcoming_minor_version):
```
```suggestion
def update_version_with_upcoming_version_in_config_py(config_py_path, upcoming_minor_version):
```

onurctirtir commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+
+def update_version_with_upcoming_version_inconfig_py(config_py_path, upcoming_minor_version):
+    print(f"### Updating {config_py_path} file with the upcoming version {upcoming_minor_version}...### ")
+    if not replace_line_in_file(config_py_path, "^MASTER_VERSION =*",
```
Can we move `^MASTER_VERSION =*` to a constant ? (and let's be more specific, by using something else instead of `*`)

In general, can we please move such hardcoded regex strings into some constants (as we did for `MULTI_EXT_MAJOR_SEARCH_PATTERN`) ?

onurctirtir replied:
minor: let's use `\.` instead of `.`

r"^MASTER_VERSION = '\d+.\d+'" -> r"^MASTER_VERSION = '\d+**\\**.\d+'"

gurkanindibay replied:
Fixed

onurctirtir commented on python/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,116 @@
+def clear_env():
+    if os.path.exists("../citus"):
+        os.chdir("..")
+        run("chmod -R 777 citus")
```
why do we need that (since we are anyway removing this directory) ?

hanefi replied:
I think it is better to clean directories and files before running the tests, not after. This way, it is easier to investigate and understand why a test fails.

gurkanindibay replied:
I think keeping environment clean is important if we clean at the start, there will be some residual files after test execution which is not a good practice. If we need to clean environment before test execution, we need to clean environment after the tests as well to remove any residual files which is unnecessary and increases the test duration

onurctirtir replied:
I agree with Hanefi on doing the cleanup before starting to do anything.
Consider this scenario: We rely on that script does the clean-up in the end.
If we interrupt the script in the middle of its execution, then we wouldn't clean-up anything.

Other than that, I think we should create a temporary directory before doing anything.
Consider below:
```python
def clear_env():
    if os.path.exists("../citus"):
        os.chdir("..")
        # run("chmod -R 777 citus")
        run("sudo rm -rf citus")
```

This function removes `../citus` directory always, no matter if it's created by the script or not.
This is dangerous a lot since that directory might be a real directory used by the person who runs the script in their local environment.

We can chat over teams if there is anything unclear.

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return filtered_prs
+
+
+def file_include_line(base_path: str, relative_file_path: str, line_content: str) -> bool:
```
Fixed

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return False
+
+
+def line_count_in_file(base_path: str, relative_file_path: str, search_line: str) -> int:
```
Fixed

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return f"{project_name}.spec"
+
+
+def get_version_number(version: str, fancy: bool, fancy_release_count: int) -> str:
```
We talked to @hanefi and we thought that we can remove it. Also for pg-auto-failover Dimitri didn't want to use fancy parameter. So to make it usable for all projects I made it parametric. We can discuss this again

onurctirtir replied:
Seems that we don't even use `get_version_number` & `get_version_number_with_project_name` in nowhere, expect the unit-tests.
I think we should directly remove those functions.

gurkanindibay replied:
Fixed

hanefi replied:
Let me make some corrections here:

PGAF already has fancy versioning. You can check out the package names. ( e.g. `-1` part in `1.5.2-1` in `postgresql-10-auto-failover-1.5_1.5.2-1_amd64.deb` )

What we discussed earlier was about the workflow parameter. Since a project will not stop using fancy versioning or start using it quite frequently, Jelte and I argued that it should be a hardcoded parameter.

We also argued that it is not necessary to have two separate variables here. However this is not as important. For example we can have a single variable `fancy_version` that when set to `-1` will imply that the project does not use fancy versioning.

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return result
+
+
+def cherry_pick_prs(prs: List[PullRequest.PullRequest]):
```
I think it gets all the commits I didn't put an extra control

onurctirtir replied:
>If merge commits should be filtered I can filter it 

Yes, I think we should skip merge commits.

hanefi replied:
When I check the old perl scripts, I do not see any logic to filter out merge commits. It may not be necessary.

See https://github.com/citusdata/tools/blob/develop/automated_packaging/prepare_release.pl

gurkanindibay replied:
I fixed it. I eliminated commits with multiple parents

onurctirtir replied:
>When I check the old perl scripts, I do not see any logic to filter out merge commits. It may not be necessary.

But including cherry-picking merge commits might bring unwanted ones too ? @hanefi

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    config_py_path = f"{exec_path}/{CONFIG_PY}"
+    distributed_dir_path = f"{exec_path}/{DISTRIBUTED_DIR_PATH}"
+
+    resource_status = ResourceStatus.INITIAL
```
I added it both testing and if required to use it further rollback scenarios. Cherry picking part is prone to errors. So rollback may be required

onurctirtir replied:
but we don't do any rollback operations at the moment, so I think we should remove that unused variable

gurkanindibay replied:
Fixed

gurkanindibay commented on python/requirements.txt:
```diff
@@ -6,7 +6,6 @@ blinker==1.4
 certifi==2019.11.28
 chardet==3.0.4
```
By using that command 
 python -m pip freeze  > requirements.txt

onurctirtir replied:
But I think that determines the `requirements.txt` according to what is installed within your environment.
For example, I could quickly find random three items in this list that we don't actually have a dependency on: Twisted, zope.interface, PyGObject

Maybe you can run `python -m pip freeze` in a fresh environment.

I think this is an important item before to merging this pr.
This is because,  in case somebody (other than you) needs to execute those scripts, they would need wait for all the things in this list to be installed, which would slow down the process a lot.

gurkanindibay replied:
Fixed using pip-tools

hanefi replied:
@gurkanindibay can you share a brief explanation on how we use `pip-tools` ? I still see the requirements.txt in the branch and I am quite confused.

gurkanindibay replied:
We still use requirements.txt . pip-tools has compile option I used pip-compile --generate-hashes option 
The steps I followwed to create requirements.txt is below
1. Create requirements.in file with required dependencies
2. create a virtual environment using pip to provide clean run
  python3 -m venv env
3. activate virtual environment
 source env/bin/activate
4. Go to directory which includes requirements.in
5. Run pip-compile with hash option
 pip-compile --generate-hashes
References:
https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/
https://github.com/jazzband/pip-tools

hanefi replied:
It would be great if you document this list of items in a readme file

gurkanindibay replied:
https://github.com/citusdata/tools/issues/104

gurkanindibay commented on python/common_validations.py:
```diff
@@ -0,0 +1,29 @@
+
+@parameter_validation
+def is_email(email: str):
+    if not string_utils.is_email(email):
```
method handles None cases no need to put extra check

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+
+from github import Github
+
+from .common_tool_methods import *
+from .common_validations import *
```
Fixed

onurctirtir replied:
We still have two occurrences of `import *` usage:

1) `from .common_validations import *`
2) `from ..update_package_properties import *`

gurkanindibay replied:
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    print(f"### OK Changes pushed for {upcoming_version_branch}### ")
+
+
+def commit_all_changes_on_branch(project_name, project_version, release_branch):
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        update_version_with_upcoming_version_in_citus_control(citus_control_file_path, upcoming_minor_version)
+        # commit and push changes on master-update-version-$curtime branch
+        commit_all_changes_on_branch(project_name, project_version, upcoming_version_branch)
+        if not is_test:
+            push_branch(upcoming_version_branch)
+            resource_status = ResourceStatus.UPCOMING_VERSION_REMOTE
+        # create pull request
+        if not is_test:
+            create_pull_request_for_upcoming_version_branch(main_branch, repository, upcoming_version,
+                                                            upcoming_version_branch)
+            resource_status = ResourceStatus.PULL_REQUEST_CREATED
```
Fixed

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return picked_pulls
+
+
+def get_prs_by_label(prs: List[PullRequest.PullRequest], label_name: str):
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+def checkout_branch(branch_name, is_test):
+    print(f"### Checking out {branch_name}...### ")
+    run(f"git checkout {branch_name}")
+    if not is_test:
```
When is_test is active we don't push branches. So there is not a remote counterpart. 
I didn't push while testing because when we push, it triggers a pipeline job which is unnecessary in this case

onurctirtir replied:
I think it makes sense to not push the branch to remote repo during tests.

On the other hand, I couldn't quite understand what does that mean ?
>When is_test is active we don't push branches. So there is not a remote counterpart.

I think there is still a remote repo from which we clone the repo.
In that sense, I think we can still do `git pull` and I think we should do.

gurkanindibay replied:
If there is not a remote counterpart, we are getting an error as below. So in testing case, since there is not such branch we would get error if we pulled
![image](https://user-images.githubusercontent.com/1596331/119642112-0f9cd000-be23-11eb-833d-c50a13452bd4.png)

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        create_branch(release_branch_name)
+        resource_status = ResourceStatus.RELEASE_BRANCH_LOCAL
+        # change version info in configure.in file
+        update_version_in_configure_in(configure_in_path, devel_version)
```
The counterpart for this line is below. It uses devel version
https://github.com/citusdata/tools/blob/c5fac60ae0fc5ec559c9c0d89fd87800841f90e0/automated_packaging/prepare_release.pl#L76

onurctirtir replied:
Yes, we should use `devel` version in `prepare_upcoming_version_branch` but we should *not* do that in `prepare_release_branch_for_major_release`. I meant, we should not use `devel_version` in below two lines:

```python
def prepare_release_branch_for_major_release(majorReleaseParams: MajorReleaseParams):
    ...
    update_version_in_configure_in(majorReleaseParams.configure_in_path, majorReleaseParams.devel_version)
    ...
    update_version_in_multi_extension_out(majorReleaseParams.multi_extension_out_path, majorReleaseParams.devel_version)
    ...
```

gurkanindibay replied:
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        content = content + "\n\n"
+        content = content + f"-- bump version to {upcoming_minor_version}-1" + "\n\n"
+        f_writer.write(content)
+    run(f"git add {distributed_dir_path}/{newly_created_sql_file}")
```
Yes it should be as you said but I had some problems in some of my tests without it

onurctirtir replied:
Can you expand what kind of problems they were ?
Ideally, as long as we know what we stage (.i.e `git add`), we shouldn't encounter with any problems.

gurkanindibay replied:
Fixed and will be tested

gurkanindibay replied:
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    newly_created_sql_file = ""
+
+    # if major release
+    if is_major_release(project_version):
```
update_release_for_major_release ->prepare_release_branch_for_major_release, prepare_upcoming_version_branch
update_release_for_patch_release->prepare_release_branch_for_patch_release 
Fixed

onurctirtir replied:
thanks for doing that

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+           resource_status
+
+
+def list_and_cherrypick_prs_with_backport_labels(earliest_pr_date, main_branch, release_branch_name, repository):
+    print(
+        f"### Getting all PR with backport label after {datetime.strftime(earliest_pr_date, '%Y.%m.%d %H:%M')}### ")
+    all_related_prs = get_prs(repository, earliest_pr_date, main_branch)
+    # get commits for selected prs with backport label
+    prs_with_backport = get_prs_by_label(all_related_prs, "backport")
+    print(f"### OK {len(prs_with_backport)} PRs with backport label found. PR list is as below### ")
+    for pr in prs_with_backport:
+        print(f"No:{pr.number} Title:{pr.title}\n")
+    # cherrypick all commits with backport label
+    print(f"Cherry-picking PRs on {release_branch_name}...")
+    cherry_pick_prs(prs_with_backport)
+    print(f"OK Cherry pick completed for all PRs on branch {release_branch_name}")
```
1. Changed the name
2. I already sort using api as below
https://github.com/citusdata/tools/blob/a956c2a02806f7474a14ad8721153ce50f24181b/python/common_tool_methods.py#L96

onurctirtir replied:
I put a more comprehensive comment for that, see https://github.com/citusdata/tools/pull/91#discussion_r638623583

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    return env
+
+
+def find_nth_overlapping(subject_string, search_string, n) -> int:
+    start = subject_string.find(search_string)
+
+    while start >= 0 and n > 1:
+        start = subject_string.find(search_string, start + 1)
+        n -= 1
+    return start
+
+
+def find_nth_overlapping_line_by_regex(subject_string, regex_pattern, n) -> int:
+    lines = subject_string.splitlines()
+    counter = 0
+    for i in range(len(lines)):
+        if re.match(regex_pattern, lines[i]):
+            counter = counter + 1
+        if counter == n:
+            return i
+    return -1
```
subject_string is str type I added str type
I didn't change the parameter name of find_nth_matching_line_number into lines because I think splitting into lines should be in this functions
Others are fixed

onurctirtir replied:
I couldn't quite understand what is happening around `find_nth_matching_line_number` & `find_nth_matching_line`.
`find_nth_matching_line` splits the lines and calls `find_nth_matching_line_number`.
However, `find_nth_matching_line_number` splits those lines too. 
I wonder how our unit tests don't fail on that.

Also a minor comment: maybe return `None` instead of empty string.

In the end, maybe we can implement an helper function and make use of it:

```python
def find_nth_matching_line(subject_string: str, regex_pattern: str, n: int) -> str:
    """Takes a subject string, regex param and the search index as parameter and returns line content of found match.
        If not found returns None"""
    lines = subject_string.splitlines()
    line_number = find_nth_matching_string_idx(subject_string, regex_pattern, n)
    if line_number != -1:
        return lines[line_number]
    else:
        return None


def find_nth_matching_line_number(subject_string: str, regex_pattern: str, n: int) -> int:
    """Takes a subject string, regex param and the search index as parameter and returns line number of found match.
        If not found returns -1"""
    lines = subject_string.splitlines()
    line_number = find_nth_matching_string_and_idx(lines, regex_pattern, n)
    return line_number


def find_nth_matching_string_idx(string_list: str, regex_pattern: str, n: int):
    counter = 0
    for idx, string in enumerate(string_list):
        if re.match(regex_pattern, string):
            counter = counter + 1
        if counter == n:
            return idx
    return -1
```

gurkanindibay replied:
I unified method to make these methods less complex and named find_nth_matching_line_and_line_number. Is it acceptable?
I usually avoid Null return types since it is not suggested as clean code principles to avoid futher null checks. But if there is a team policy I will obey this

hanefi replied:
I don't think we have a general rule for this. Each project we maintain have slightly different set of principles.

Personally I am fine with this piece of code either way.

gurkanindibay replied:
By the way "do not return Null" value is a rule in "Clean Code" book If I can find, I will share it with you

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+        commits = pr.get_commits()
+        for single_commit in commits:
+            cp_result = run(f"git cherry-pick {single_commit.commit.sha}")
+            print(f"Cherry pick result for PR No {pr.number} and commit sha {single_commit.commit.sha}: {cp_result}  ")
```
Fixed 
Searched all code base to find extra whitespace while printint I haven't encountered another one

onurctirtir replied:
I still see occurrences of such whitespaces, like:

`Executing Major release flow### ")`
`Executing Patch release flow ### ")`

I think we should search & replace the following:
`[whitespace]'` -> `'`
`[whitespace]"` -> `"`
`'[whitespace]` -> `'`
`"[whitespace]` -> `"`

gurkanindibay replied:
Fixed 
As a note I'm sharing the regex I used for searching
print\(f*\'.* \'\)
print\(f*\".* \"\)

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        if not is_test:
+            push_branch(release_branch_name)
+            resource_status = ResourceStatus.RELEASE_BRANCH_REMOTE
+        print(f"### OK {release_branch_name} prepared.### ")
```
But not just pushed. I mean all the process. I don't push the code if is_test true as well. So pushed woulld not be a correct description

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+            push_branch(release_branch_name)
+            resource_status = ResourceStatus.RELEASE_BRANCH_REMOTE
+        print(f"### OK {release_branch_name} prepared.### ")
+        # Increase version number and change upcoming version
+
```
Removed comment since method name is descriptive enough. I created two methods
prepare_release_branch_for_major_release
prepare_upcoming_version_branch

onurctirtir replied:
thanks for doing that

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        print(f"### OK {release_branch_name} prepared.### ")
+        # Increase version number and change upcoming version
+
+        print(f"### Preparing upcoming version branch for the new master with name {upcoming_version_branch}...### ")
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        update_schema_version_with_upcoming_version_in_multi_extension_file(current_schema_version,
+                                                                            multi_extension_sql_path,
+                                                                            upcoming_minor_version, upcoming_version)
+        # find current schema version info and update it with upcoming version  in multi_extension.out file
```
Searched with regex and found and fixed them

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        update_schema_version_with_upcoming_version_in_multi_extension_file(current_schema_version,
+                                                                            multi_extension_out_path,
+                                                                            upcoming_minor_version, upcoming_version)
+        # create new sql file with the name ">./src/backend/distributed/citus--$current_schema_version--
+        # $upcoming_minor_version-1.sql"
```
Fixed

onurctirtir replied:
can you open an issue for that, so we don't forget doing that ?:

>Indeed, we also need to put a similar file for downgrade path but I think we should work on that in a separate pr, in future.

gurkanindibay replied:
https://github.com/citusdata/tools/issues/103

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+            resource_status = ResourceStatus.PULL_REQUEST_CREATED
+    else:
+        print(f"### {project_version} is a patch release. Executing Patch release flow ### ")
+        # checkout release branch in release-X.Y format
```
Fixed

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+            continue
+        if pull.created_at < earliest_date:
+            break
+        if pull.merged_at is not None and earliest_date < pull.merged_at < last_date:
+            picked_pulls.append(pull)
```
Fixed

gurkanindibay replied:
Fixed as suggested in https://github.com/citusdata/tools/pull/91#discussion_r638623583

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+
+        # change version info in configure.in file
+        update_version_in_configure_in(configure_in_path, project_version)
+        # execute "auto-conf " (Not with f flag!!!)
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+        update_version_in_multi_extension_out(multi_extension_out_path, project_version)
+
+        if cherry_pick_enabled:
+            # list all pr's with backport labels
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+            # list all pr's with backport labels
+            list_and_cherrypick_prs_with_backport_labels(earliest_pr_date, main_branch, release_branch_name, repository)
+
+        # commit all-changes
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+def list_and_cherrypick_prs_with_backport_labels(earliest_pr_date, main_branch, release_branch_name, repository):
+    print(
+        f"### Getting all PR with backport label after {datetime.strftime(earliest_pr_date, '%Y.%m.%d %H:%M')}### ")
+    all_related_prs = get_prs(repository, earliest_pr_date, main_branch)
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    for pr in prs_with_backport:
+        print(f"No:{pr.number} Title:{pr.title}\n")
+    # cherrypick all commits with backport label
+    print(f"Cherry-picking PRs on {release_branch_name}...")
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+
+
+def create_branch(release_branch_name):
+    print(f"### Creating release branch with name {release_branch_name}...### ")
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    print(f"### Updating version on file {configure_in_path}...### ")
+    if not replace_line_in_file(configure_in_path, CONFIGURE_IN_SEARCH_PATTERN,
+                                f"AC_INIT([Citus], [{project_version}])"):
+        raise ValueError(f"{configure_in_path} does not have match for version")
```
Fixed

gurkanindibay commented on python/prepare_release.py:
```diff
@@ -0,0 +1,296 @@
+    print(f"### OK Current schema version updated on {multi_extension_sql_path} to {upcoming_minor_version}### ")
+
+
+def get_current_schema_from_citus_control(citus_control_file_path):
+    print(f"### Reading current schema version from  {citus_control_file_path}...### ")
+    current_schema_version = get_current_schema_version(citus_control_file_path)
+    if len(current_schema_version) == 0:
+        raise ValueError("Version info could not be found in citus.control file")
+    print(f"### OK Current schema version is {current_schema_version}### ")
+    return current_schema_version
```
1. Unified methods
2. crete a new method to read the nth line  and used it

gurkanindibay commented on python/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,116 @@
+        clear_env()
+
+
+def clear_branches(release_branch, resource_status, upcoming_version_branch):
```
I removed it since it is not required nor. Because we are not pushin in tests. There is no need to cleatr branches because at the end of tests we already delete cloned code

onurctirtir replied:
makes a lot sense

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+    for pull in pulls:
+        if not pull.is_merged():
+            continue
+        if pull.created_at < earliest_date:
+            break
```
I changed the logic . I added a filter logic instead of while loop. Now it is leaner

onurctirtir replied:
1) seems that we still compare created_at with something.
I don't think we need to check creation date, but the merge date.

2) We will anyway check if pr is merged, but I think we can use `state='closed'` instead of 'all'

3) List comprehension is a nice python feature but I think having that much if's in an in-line expression decreases the readability.

4) Looking into our previous perl script, passing `sort='created'` doesn't seem to be enough.
I think we need to sort the pr's by their merge date. On the other hand, github api doesn't seem to be providing that

In the end, I would imagine this function as such:
```python
def get_prs_for_patch_release(repo: Repository.Repository, earliest_date: datetime, base_branch: str, last_date: datetime = None):
    pull_requests = repo.get_pulls(state="closed", base=base_branch, sort="created", direction="desc")

    # filter pull requests according to given time interval
    filtered_pull_requests = list()
    for pull_request in pull_requests:
        if not pull_request.is_merged():
            continue
        if p.merged_at < earliest_date:
            continue
        if last_date and p.merged_at > last_date:
            continue

        filtered_pull_requests.append(pull_request)

    # finally, sort the pr's by their merge date
    sorted_pull_requests = sorted(filtered_pull_requests,
                                  key=lambda pull_request: pull_request.merged_at)
    return sorted_pull_requests
```

gurkanindibay replied:
Fixed and willl be tested

gurkanindibay replied:
Fixed

gurkanindibay commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+def get_default_upcoming_version(version: is_version(str)) -> str:
+    project_version_details = get_version_details(version)
+    return f'{project_version_details["major"]}.{project_version_details["minor"]}.' \
+           f'{str(int(project_version_details["patch"]) + 1)}'
```
Fixed

hanefi commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,341 @@
+        push_branch(release_pr_branch)
+
+
+def prepare_upcoming_version_branch(citus_control_file_path, config_py_path, configure_in_path, devel_version,
```
So many parameters that are hard to keep track. Can we pass them in a param object?

gurkanindibay replied:
Fixed

hanefi commented on packaging_automation/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,106 @@
+
+def clear_env():
+    if os.path.exists("../citus"):
+        os.chdir("..")
+        run("chmod -R 777 citus")
+        run("sudo rm -rf citus")
```
Why do we need to set file modes before deleting them?

```suggestion
        run("sudo rm -rf ../citus")
```

gurkanindibay replied:
Fixed

hanefi replied:
I do not agree that this one is actually fixed yet. We still have one extra line.

hanefi replied:
See https://github.com/citusdata/tools/pull/91/files#r641363744 for my suggestion in the renamed file.

gurkanindibay replied:
Fixed

hanefi commented on packaging_automation/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,106 @@
+            project_name="citus",
+            project_version="10.2.0",
+            main_branch=MAIN_BRANCH,
+            earliest_pr_date=datetime.strptime(
+                '2021.03.25 00:00',
+                '%Y.%m.%d %H:%M'),
```
```suggestion
            earliest_pr_date=datetime.strptime('2021.03.25 00:00',
                	                           '%Y.%m.%d %H:%M'),
```

gurkanindibay replied:
Fixed

gurkanindibay commented on packaging_automation/publish-packages.py:
```diff
@@ -0,0 +1,103 @@
+from pprint import pprint
+
+ms_package_repo_map = {
+    "el/8": "centos-8", "el/7": "centos-7", "debian/buster": "debian-buster", "debian/jessie": "debian-jessie",
```
It makes the file harder to read. Because it will take too much horizontal space in your suggested case which will require lots of scrolling

onurctirtir replied:
For readability, I agree with Hanefi on that

hanefi replied:
@gurkanindibay can you check if you actually fixed this? You renamed the file and it still has the same contents here as before

gurkanindibay replied:
I removed pusblish-packages to open another PR for that

gurkanindibay commented on bash/update_package_properties.sh:
```diff
@@ -26,7 +26,7 @@ commit_message="Bump to ${PRJ_NAME} ${PRJ_VER}"
 git checkout -b "${pr_branch_name}"
 
-python tools/python/update_package_properties.py --gh_token "${GH_TOKEN}" --prj_name "${PRJ_NAME}" --prj_ver "${PRJ_VER}" --tag_name "${TAG_NAME}" --fancy "${FANCY}" \
+python tools/python/update_package_properties.py --gh_token "${GH_TOKEN}" --prj_name "${PRJ_NAME}"  --tag_name "${TAG_NAME}" --fancy "${FANCY}" \
```
Fixed

gurkanindibay commented on .github/workflows/tool-tests.yml:
```diff
@@ -25,10 +25,14 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v2
+      - name: Define git credentails
+        run: git config --global user.email "gindibay@microsoft.com"&& git config --global user.name "Gurkan Indibay"
```
Fixed

gurkanindibay commented on .github/workflows/tool-tests.yml:
```diff
@@ -25,10 +25,14 @@ jobs:
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
+      - name: Define git credentails
```
Fixed

onurctirtir commented on python/common_tool_methods.py:
```diff
@@ -0,0 +1,146 @@
+
+def get_upcoming_minor_version(version: is_version(str)) -> str:
+    upcoming_version_details = get_version_details(version)
+    return f'{upcoming_version_details["major"]}.{upcoming_version_details["minor"]}'
```
second item seems to be not resolved yet

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,395 @@
+
+
+def create_and_checkout_branch(release_branch_name):
+    get_current_branch()
```
I think we can remove that

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,395 @@
+        schema_not_found = False
+        if len(cc_line) > 0:
+            line_parts = cc_line.split("=")
+            if len(line_parts) > 0:
```
let's be more explicit:

```suggestion
            if len(line_parts) == 2:
```

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,100 @@
+def clear_env():
+    if os.path.exists("../citus"):
+        os.chdir("..")
+        # run("chmod -R 777 citus")
```
I think we can remove that commented line

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def remove_string_inside_parentheses(param: str) -> str:
+    return re.sub(r"[(\[].*?[)\]]", "", param)
```
1) seems that this removes the parenthesis too (since the second argument is `""`)

2) parentheses -> parenthesis

3) putting an example for that via comment would be useful: # e.g: [string] -> []

gurkanindibay replied:
Changed method name into remove_text_with_parenthesis
Put docstring onto method

onurctirtir replied:
two minor comments:

* Removes texts within parenthesis together with the parenthesis itself ..
* parentheses -> parenthesis

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/publish-packages.py:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+
```
btw, can we please add `publish-packages.py` script in another pr ? since it doesn't seem be related with prepare release phase

onurctirtir reviewed (commented):
I think we are pretty close to merging this pr.

Other than a few new comments, I put some additional comments to unresolved ones from my earlier pass (please click `Load more...` buttons).

I will do manual testing after all the previous & new comments left by me & Hanefi are addressed, since some of them are behavioral changes.

onurctirtir commented on packaging_automation/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,100 @@
+                       earliest_pr_date=datetime.strptime('2021.03.25 00:00', '%Y.%m.%d %H:%M'),
+                       exec_path=TEST_BASE_PATH, is_test=True)
+
+        update_release_return_value = update_release(
```
`update_release_return_value`  seems to be unused

gurkanindibay replied:
Fixed

hanefi commented on packaging_automation/tests/test_prepare_release.py:
```diff
@@ -0,0 +1,100 @@
+
+def clear_env():
+    if os.path.exists(f"../{TEST_CHECKOUT_DIR}"):
+        os.chdir("..")
+        run(f"sudo rm -rf {TEST_CHECKOUT_DIR}")
```
```suggestion
        run(f"sudo rm -rf ../{TEST_CHECKOUT_DIR}")
```

gurkanindibay replied:
I need to go out of the directory then I can delete it otherwise I got exception because I'm still in the directory that I've deleted. Unit tests for patch release failed so I rollbacked the development

hanefi reviewed (commented):
I left some more comments, and I am fine with merging after making sure all open items are addressed.

One thing that bothers me is the inconsistent output formats.
- Please have one space before closing `###`
- Consider using `...` when printing the outputs before doing something that takes time to complete. Preferably with a `done\n` after it completes. This leads to the most intuitive user experience IMO.

Please go over the all the log messages.

```
bash$ grep -rho 'f"###.*###"' . | sort
f"### Checking out {branch_name}...###"
f"### Committing changes for branch {current_branch}...###"
f"### Creating file {newly_created_sql_file}...###"
f"### Creating pull request for {upcoming_version_branch}###"
f"### Creating release branch with name {release_branch_name} from {get_current_branch()}...###"
f"### Executing autoconf -f command...###"
f"### OK Changes committed for {current_branch}###"
f"### OK Changes pushed for {upcoming_version_branch}###"
f"### OK Current schema version updated on {multi_extension_sql_path} to {upcoming_minor_version}###"
f"### OK Pull request created. PR Number:{pr_result.number} PR URL: {pr_result.url}###"
f"### OK autoconf -f executed.###"
f"### OK {branch_name} checked out and pulled###"
f"### OK {citus_control_file_path} file is updated with the upcoming version {upcoming_minor_version}...###"
f"### OK {configure_in_path} file is updated with project version {project_version}.###"
f"### OK {len(prs_with_backport)} PRs with backport label found. PR list is as below###"
f"### OK {majorReleaseParams.release_branch_name} prepared.###"
f"### OK {multi_extension_out_path} file is updated with project version {project_version}.###"
f"### OK {multi_extension_out_path} file is updated with project version {project_version}.###"
f"### OK {release_branch_name} created###"
f"### OK {upcoming_params.upcoming_version_branch} prepared.###"
f"### Preparing {majorReleaseParams.release_branch_name}...###"
f"### Preparing {upcoming_params.upcoming_version_branch} branch that bumps master version.###"
f"### Reading current schema version from {citus_control_file_path}...###"
f"### Updating version on file {configure_in_path}...###"
f"### Updating {citus_control_file_path} file with the upcoming version {upcoming_minor_version}...###"
f"### Updating {config_py_path} file with the upcoming version {upcoming_minor_version}...###"
f"### Updating {multi_extension_out_path} file with the project version {project_version}...###"
f"### Updating {multi_extension_out_path} file with the project version {project_version}...###"
f"### {config_py_path} file updated with the upcoming version {upcoming_minor_version}...###"
f"### {patchReleaseParams.project_version} is a patch release. Executing Patch release flow ###"
f"### {project_version} is a major release. Executing Major release flow###"
```

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def remove_text_with_parenthesis(param: str) -> str:
+    """Removes texts within parentheses i.e. outside parenthesis(inside parenthesis)-> outside parenthesis """
```
two minor comments:

* Removes texts within parenthesis **together with the parenthesis itself** ..
* parentheses -> parenthesis

gurkanindibay commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,394 @@
+    repository = g.get_repo(f"{REPO_OWNER}/{project_name}")
+    newly_created_sql_file = ""
+
+    # if major release
+    if is_major_release(project_version):
```
Fixed

gurkanindibay commented on packaging_automation/tests/test_common_tool_methods.py:
```diff
@@ -0,0 +1,92 @@
+        result = run("echo 'Run' method is performing fine ")
+        self.assertEqual(0, result.returncode)
+
+    def test_remove_parentheses_from_string(self):
```
Fixed

gurkanindibay commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,394 @@
+        content = content + "\n\n"
+        content = content + f"-- bump version to {upcoming_minor_version}-1" + "\n\n"
+        f_writer.write(content)
+    # run(f"git add {distributed_dir_path}/{newly_created_sql_file}")
```
Fixed

gurkanindibay replied:
I went over all messages. I standardized all message formats

hanefi commented on packaging_automation/prepare_release.py:
```diff
@@ -0,0 +1,394 @@
+
+        newly_created_sql_file = prepare_upcoming_version_branch(branch_params)
+        print(f"OK {project_version} Major release flow executed successfully")
+    else:
```
I do not see a comment here, so how is my comment above fixed exactly? See https://github.com/citusdata/tools/pull/91/files#r641466250

gurkanindibay replied:
I haven't pushed yet. First I write Fixed then I will push it in one commit. You are too fast for me :)

hanefi reviewed (approved):
I am ok with the changes in this PR

Please organize the commits before merging and make sure you get an approval from @onurctirtir as well.

onurctirtir commented on .github/workflows/tool-tests.yml:
```diff
@@ -2,6 +2,9 @@ name: Tool Tests
 
```
Sharing some additional comments from my manual testing:

* Minor release flow seems pretty fine.

* Major release flow: 
Release branch creation works fine.
But the changes made for master branch doesn't seem to be correct. Please check with the commit that I shared before

* Getting some errors after running the script for the second time:

```
....
....
subprocess.CalledProcessError: Command 'git checkout -b release-10.2-test' returned non-zero exit status 128.
```

* For the missing/incorrect parameters, I would expect throwing errors beforehand, but for the most, we throw the error in very deeper levels (like when using github api or when splitting a string etc.).
Let's make some correctness checks for them at the very beginning of our main entrypoint.
I think we can have a function to do those checks such as `check_args` or something.

* In `initialize_env` function, we should decide which repo to clone according to `prj_name` arg. 
It can be either "citus" or "citus-enterprise".  So I think we can use `add_arg(..choices=..)` for that.

* For `--is_test` & `cherry_pick_enabled` params, I think we can use `add_arg(..action='store_true'..`, instead of requiring to pass "True" string.

* For `schema_version`, I think we should error if it's passed for major release. 
But we should be okay if it is passed or not passed for patch release.
 
* I think we should set `add_arg(..required=true`..), for `prj_ver`, `main_branch` & `gh_token`.

* We should error if either of `cherry_pick_enabled` or `earliest_pr_date` is set for major release.
But we should be okay if it is passed or not passed for patch release.

* We should error if `cherry_pick_enabled` is set to true but `earliest_pr_date` is not given.

* Can we simplify `earliest_pr_date` format, like just `'%Y.%m.%d` ? I don't think we need to know hour, minute etc. ever.

gurkanindibay replied:
I addressed all of them except the action="store_true" parameters because when this parameter is used, it can not be set by user. It sets a constant variable not suitable for user-defined parameters

onurctirtir replied:
> it can not be set by user.

Assuming `is_test` arg, using `store_true` means:
If user doesn't add `--is_test`, then `is_test` would automatically evaluate to `false`, so we will disable test mode.
If user puts `--is_test` into the args (without any additional strings), then `argparese` should interpret `is_test` as `true`.

https://stackoverflow.com/a/8203679/13370386

Is there any reason for not using that ?

gurkanindibay replied:
I hadn't totally grasped the parameter. As you referenced, I understood and changed that way thanks 👍

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,286 @@
+BASE_GIT_PATH = pathlib2.Path(__file__).parents[1]
+PATCH_VERSION_MATCH_FROM_MINOR_SUFFIX = "\.\d{1,3}"
+
+# When using GitPython library Repo objects should be closed to be able to delete cloned sources
+# referenced by Repo objects. References are stored below array to be able to close all resources after code execution
```
Minor:

```suggestion
# When using GitPython library, Repo objects should be closed to be able to delete cloned sources
# referenced by Repo objects. References are stored in below array to be able to close all resources after code execution.
```

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,286 @@
+    return has_match
+
+
+def append_line_in_file(file: str, match_regex: str, append_str: str) -> bool:
```
I couldn't quite understand what we are exactly doing in `append_line_in_file` & `prepend_line_in_file`.

Can you improve the readability of those two functions, or at least let's comment why do we do `anIndex +1` / `anIndex +2` kind of things.

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,286 @@
+    return repo.active_branch
+
+
+def does_remote_branch_exist(branch_name: str, working_dir: str) -> bool:
```
minor:

`does_` prefix sounds a bit unusual to me, maybe:

does_remote_branch_exist -> remote_branch_exists
does_local_branch_exist -> local_branch_exists
does_branch_exist -> branch_exists

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,286 @@
+        if os.path.exists(f"{exec_path}/.git"):
+            run(f"chmod -R 777 {exec_path}/.git")
+        try:
+            run(f"sudo rm -rf {exec_path}")
```
Why do we use `sudo` here ? Can we please remove that ?

gurkanindibay replied:
Fixed

onurctirtir commented on packaging_automation/common_tool_methods.py:
```diff
@@ -0,0 +1,286 @@
+        commits = pr.get_commits()
+        for single_commit in commits:
+            if not is_merge_commit(single_commit):
+                cp_result = run(f"git cherry-pick {single_commit.commit.sha}")
```
We usually use `-x` parameter to cherry-pick the commits:

```
       -x
           When recording the commit, append a line that says "(cherry picked from commit ...)" to the original
           commit message in order to indicate which commit this change was cherry-picked from. ...
```

```suggestion
                cp_result = run(f"git cherry-pick -x {single_commit.commit.sha}")
```

gurkanindibay replied:
Fixed

onurctirtir reviewed (commented):
Works fine for community releases, but there are some problems with enterprise releases.
Please fix those problems by comparing the example commits that I shared via Teams chat.

---------

Also, please add a good readme on how to use this script. Sharing an example in below, let's use a nicer wording:
```
(file name is something like README-prepare-release / README-prepare-release.txt)

First make sure that you have the dependencies for our script:
< one liner instruction to install dependencies >

For both major & patch release, provide `--prj-name` either as "citus" or "citus-enterprise".
Also provide `--main_branch` as `master` or `enterprise-master` depending on --prj-name` param.

For major release, do:
python3.x something --some params -- other params --an example version

For patch release, do:
python3.x something --some params -- other params --an example version

If you would like to bump the schema version for the patch release, also add X param like:
python3.x something --some params -- other params --an example version

If you would like to cherry pick commits for a patch release, also add Y, X .. params like:
python3.x something --some params -- other params --an example version
```

onurctirtir commented on packaging_automation/README.md:
````diff
@@ -0,0 +1,79 @@
+```
+#### Patch
+``` console
+python -m  packaging_automation.prepare_release --gh_token <your-personal-token> --prj_name citus-enterprise --prj_ver 10.0.4 --schema_version 10.0-5
````
let's also add here `--main-branch enterprise-master`

gurkanindibay replied:
There is no need to add main-branch because main branch is set by project_name if not given as parameter
https://github.com/citusdata/tools/blob/25ff7c1e1c5349a60ab2490d2a3187488f7fadcf/packaging_automation/prepare_release.py#L563

onurctirtir commented on packaging_automation/README.md:
```diff
@@ -0,0 +1,79 @@
+**--prj_ver:** Upcoming version to be used for release. should include three level of digits separated by dots, e.g: 10.0.1
+(Required)
+
+**--main-branch:** Branch to be used as base to be used for configuration changes. There is no need for base scenario. 
```
should be `--main_branch` according to code, not `--main-branch`

gurkanindibay replied:
Fixed

onurctirtir reviewed (approved):
Looks pretty much fine to me, thank you for all the work & addressing the comments.
Please follow `squash` strategy to merge your pr.

Left some minor final notes:

1) I run the script a few times and only in one of them it asked for `sudo`, not in the others.
Might worth to make sure that we fixed that issue.


2) Seeing three files like `test1e58b070-b0bb-4f78-9ae4-3672bd131097`, do we ever use them (and why) ? If not, can we remove them ?

3) For patch releases, we create two (release) test branches, can we avoid creating the later one (the one with uuid suffix) ?:
```
  release-10.0-test
  release-10.0-test_da502a7a-109e-44d5-92c6-d042e095b7e0
```

gurkanindibay commented:
> Looks pretty much fine to me, thank you for all the work & addressing the comments.
> Please follow `squash` strategy to merge your pr.
> 
> Left some minor final notes:
> 
> 1. I run the script a few times and only in one of them it asked for `sudo`, not in the others.
>    Might worth to make sure that we fixed that issue.
> 2. Seeing three files like `test1e58b070-b0bb-4f78-9ae4-3672bd131097`, do we ever use them (and why) ? If not, can we remove them ?
> 3. For patch releases, we create two (release) test branches, can we avoid creating the later one (the one with uuid suffix) ?:
> 
> ```
>   release-10.0-test
>   release-10.0-test_da502a7a-109e-44d5-92c6-d042e095b7e0
> ```

1. I haven't experiences such an issue. I opened an issue about it. https://github.com/citusdata/tools/issues/106
2. Fixed
3. The latter one with uuid is the one being used for PR. The other one with test is the release branch for test. If test is not used this branch will not be created. This branch is for testing purposes not to break release branches
